pkg/metricsadapter/controller.go

```diff
@@ -0,0 +1,158 @@
+	"github.com/karmada-io/karmada/pkg/util/gclient"
+)
+
+// MetricsController is a controller for metrics, control the lifecycle of multi-clusters informer
+type MetricsController struct {
+	InformerFactory informerfactory.SharedInformerFactory
+	ClusterLister   clusterlister.ClusterLister
+	InformerManager genericmanager.MultiClusterInformerManager
+
+	queue      workqueue.RateLimitingInterface
+	restConfig *rest.Config
+}
+
+// NewMetricsController creates a new metrics controller
+func NewMetricsController(restConfig *rest.Config, factory informerfactory.SharedInformerFactory) *MetricsController {
+	clusterLister := factory.Cluster().V1alpha1().Clusters().Lister()
+	controller := &MetricsController{
+		InformerFactory: factory,
+		ClusterLister:   clusterLister,
+		InformerManager: genericmanager.GetInstance(),
+		restConfig:      restConfig,
+		queue:           workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "metrics-adapter"),
+	}
+	controller.addEventHandler()
+
+	return controller
+}
+
+// addEventHandler adds event handler for cluster
+func (m *MetricsController) addEventHandler() {
+	clusterInformer := m.InformerFactory.Cluster().V1alpha1().Clusters().Informer()
+	_, err := clusterInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: m.addCluster,
+		// Update event and delete event will be handled by the same handler
+		UpdateFunc: m.updateCluster,
+	})
+	if err != nil {
+		klog.Errorf("Failed to add cluster event handler for cluster: %v", err)
+	}
+}
+
+// addCluster adds cluster to queue
+func (m *MetricsController) addCluster(obj interface{}) {
+	cluster := obj.(*clusterV1alpha1.Cluster)
+	m.queue.Add(cluster.GetName())
+}
+
+// updateCluster updates cluster in queue
+func (m *MetricsController) updateCluster(oldObj, curObj interface{}) {
+	curCluster := curObj.(*clusterV1alpha1.Cluster)
+	oldCluster := oldObj.(*clusterV1alpha1.Cluster)
+	if curCluster.ResourceVersion == oldCluster.ResourceVersion {
+		// no change, do nothing.
+		return
+	}
+
+	if oldCluster.DeletionTimestamp.IsZero() != curCluster.DeletionTimestamp.IsZero() {
+		// cluster is being deleted.
+		m.queue.Add(curCluster.GetName())
+	}
+
+	if util.ClusterAccessCredentialChanged(curCluster.Spec, oldCluster.Spec) ||
+		util.IsClusterReady(&curCluster.Status) != util.IsClusterReady(&oldCluster.Status) {
+		// Cluster.Spec or Cluster health state is changed, rebuild informer.
+		m.InformerManager.Stop(curCluster.GetName())
+		m.queue.Add(curCluster.GetName())
+	}
+}
+
+// startController starts controller
+func (m *MetricsController) startController(stopCh <-chan struct{}) {
+	m.InformerFactory.WaitForCacheSync(stopCh)
+
+	go wait.Until(m.worker, time.Second, stopCh)
+
+	go func() {
+		<-stopCh
+		genericmanager.StopInstance()
+		klog.Infof("Shutting down karmada-metrics-adapter")
+	}()
+}
+
+// worker is a worker for handle the data in queue
+func (m *MetricsController) worker() {
+	for m.handleClusters() {
```

Comment:
Can we use `util.Worker` mode ？ i.e https://github.com/karmada-io/karmada/blob/master/pkg/scheduler/scheduler.go#L224-L227

Reply:
This logic is copied from `karmada-search`, which looks almost the same as `util.Worker`. I prefer to keep it the same as search.